synapse/storage/databases/main/events_worker.py

```diff
@@ -564,10 +558,16 @@ async def _get_events_from_cache_or_db(
 
                 event_entry_map.update(missing_events)
             except Exception as e:
-                fetching_deferred.errback(e)
+                with PreserveLoggingContext():
+                    fetching_deferred.errback(e)
                 raise e
+            finally:
+                # Ensure that we mark these events as no longer being fetched.
+                for event_id in missing_events_ids:
+                    self._current_event_fetches.pop(event_id, None)
```

Comment:
I wonder if doing this after  `fetching_deferred.errback` could cause races. I can't really think how it could, though.

Reply:
I think the only thing that can happen is that another request waits on the deferred after its already been resolved, which shouldn't be an issue as awaiting on the deferred will just return immediately?